### PR TITLE
Implement aten::index | feat(torchlib) (#862)

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #883

---

**This change implements the logic for `aten::index` and adds tests for different nd index combinations and permutations.**

## Understanding `aten::index`

For `arg0` with shape `[7, 3, 4, 5, 6]`
The indexing operation `arg0[0, :, 1:2, tensor([[4,5]])]` will be translated to
```
+>  select: i64[3, 4, 5, 6] = torch.ops.aten.select.int(arg0, 0, 0);
+>  slice_1: i64[3, 4, 5, 6] = torch.ops.aten.slice.Tensor(select, 0, 0, 9223372036854775807);
+>  slice_2: i64[3, 1, 5, 6] = torch.ops.aten.slice.Tensor(slice_1, 1, 1, 2);
+>  index: i64[3, 1, 1, 2, 6] = torch.ops.aten.index.Tensor(slice_2, [None, None, arg1]);
```
Here,
- `indices = [None, None, arg1]` is equivalent to `indices = [None, None, arg1, None]`
- The operation `arg0[0, :, 1:2, tensor([[4,5]])]` is equivalent to `arg0[0, :, 1:2, tensor([[4,5]]), :]`
None in `indices` are like fillers for dimensions that cannot be removed in the process.

## Gather op reference

- https://github.com/openxla/xla/blob/main/docs/operation_semantics.md?rgh-link-date=2023-07-13T01%3A09%3A16Z#gather
- https://www.pathpartnertech.com/gather-scatter-operation-in-deep-learning-framework/

---------

Co-authored-by: BowenBao <bowbao@microsoft.com>

### CODE REVIEW

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py [hunk 1]
```diff
@@ -473,6 +471,8 @@ def _capture_graph_and_evaluate_torch_script_evaluator(function: Callable, args,
                         input.value = subarg
                         sequence_input.append(input)
                         ort_inputs[input_name] = subarg
+                    else:
+                        sequence_input.append(subarg)
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/ops_test_common.py:475]: Let's put comment here explaining why this is needed in case we forget?

**Review by BowenBao — APPROVED** [review]
🎉 LGTM. Do you plan to add bool mask indices support in this or another PR?

**Comment by justinchuby** [comment]
> 🎉 LGTM. Do you plan to add bool mask indices support in this or another PR?

I plan to do it in another PR to keep this one manageable.

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py [hunk 2]
```diff
@@ -276,10 +277,7 @@ def convert_tensor_to_numpy(input: Any) -> Any:
 
 
 def convert_kwargs_for_onnx(kwargs: dict[str, Any]) -> dict[str, Any]:
-    """Converts kwargs to be compatible with ONNX Runtime.
-
-    ONNX Runtime doesn't support torch.bool, so we convert them to torch.uint8.
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test_common.py:281]: Copilot produced something that was false

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:475]: Done

File: onnxscript/tests/function_libs/torch_lib/extra_opinfo.py [hunk 3]
```diff
@@ -444,6 +444,49 @@ def sample_inputs_col2im(op_info, device, dtype, requires_grad, **kwargs):
...
+    del op_info  # Unused
+    del kwargs  # Unused
+    make_arg = functools.partial(
+        torch_testing.make_tensor, dtype=dtype, device=device, requires_grad=requires_grad
+    )
+    s = 5
+    index_1d = common_methods_invocations.index_variable(2, s, device=device)
+    index_2d = common_methods_invocations.index_variable((s + 1, 2), s, device=device)
+    index_3d = common_methods_invocations.index_variable((s + 2, s + 1, 2), s, device=device)
+    test_args = [
```
Comment (thiagocrepaldi) [inline onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:457]: would itertools.{product,permutation,combination} help in ensuring all combinations are covered and make the code shorter?

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 2 above)
Comment (thiagocrepaldi) [inline onnxscript/tests/function_libs/torch_lib/ops_test_common.py:281]: for my knowledge, does ort support torch.bool now? or was this docstring outdated already?

File: onnxscript/tests/function_libs/torch_lib/extra_opinfo.py (hunk 3 above)
Comment (thiagocrepaldi) [inline onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:457]: could `itertools.product` (or friends) with length 1 to 4 could shorten this listen and ensure no combination is left out?

**Review by thiagocrepaldi — APPROVED** [review]
LGTM

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:281]: I actually don't know if ORT supports bool (it should?), but I think this message was a mistake by copilot because we don't actually have this conversion logic. If we see issues with ORT I will make adjustments.

File: onnxscript/tests/function_libs/torch_lib/extra_opinfo.py (hunk 3 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:457]: Good point. Let me try that

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:457]: So I tested with itertools but realized some combinations are invalid so we cannot enumerate them all using itertools. For the sake of clarity I propose that we keep the current explicit tests. I also added more test cases and comments

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:457]: As above. Turns out some combinations are invalid to torch and it may be better to specify explicitly.

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:477]: Good point. I wonder if there are things we don't expect that can sneak in? Nested lists?

Comment (titaiwangms) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:477]: No idea. Just in case.

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:477]: SG. Created https://github.com/microsoft/onnxscript/issues/892 for now.
